File: backend/src/modules/transportation/suppliers/suppliers.controller.ts
import { Body, Controller, Get, Headers, Post, UseGuards } from '@nestjs/common';
import { HeaderAuthGuard, Validation } from '../../../common/auth/header-auth.guard';
import { CurrentUser } from '../../../common/auth/current-user.decorator';
import { SuppliersService } from './suppliers.service';
import { decodeHeader } from '../../../common/http/header.util';

// Base path /api/Transportation (faithful to the documented URLs); every action
// requires the CompanyName + UserToken headers via the guard. Filters/pagination
// ride in HTTP headers (Node lowercases them), bodies are JSON — as documented (§6.12).
@Controller('api/Transportation')
@UseGuards(HeaderAuthGuard)
export class SuppliersController {
  constructor(private readonly service: SuppliersService) {}

  // NOTE: not in API.md — needed by the suppliers screen (list + create).
  @Get('getSuppliers')
  getSuppliers(
    @Headers('pageno') pageNo?: string,
    @Headers('noofitems') noOfItems?: string,
    @Headers('name') name?: string,
    @Headers('phone') phone?: string,
    @Headers('mobile') mobile?: string,
  ) {
    return this.service.getSuppliers(Number(pageNo) || 1, Number(noOfItems) || 20, {
      name: decodeHeader(name),
      phone: phone || undefined,
      mobile: mobile || undefined,
    });
  }

  @Post('addSupplier')
  addSupplier(@Body() body: SupplierBody) {
    return this.service.addSupplier(body);
  }

  // Single-supplier read/update (closes the frontend's mock fallback for the
  // supplier detail page). Not in API.md — added.
  @Get('getSupplier')
  getSupplier(@Headers('id') id?: string) {
    return this.service.getSupplier(Number(id));
  }

  @Post('updateSupplier')
  updateSupplier(@Body() body: SupplierBody & { Id: number }) {
    return this.service.updateSupplier(Number(body?.Id), body);
  }

  @Get('AccountsAllMonthsForSupplier')
  accountsAllMonths(
    @Headers('year') year?: string,
    @Headers('supplierid') supplierId?: string,
    @Headers('routeid') routeId?: string,
    @Headers('month') month?: string,
    @Headers('pageno') pageNo?: string,
    @Headers('noofitems') noOfItems?: string,
  ) {
    return this.service.accountsAllMonths(Number(pageNo) || 1, Number(noOfItems) || 20, {
      year: Number(year) || undefined,
      supplierId: Number(supplierId) || undefined,
      routeId: Number(routeId) || undefined,
      month: Number(month) || undefined,
    });
  }

  @Get('AccountsAllMonthsForSupplierExcel')
  accountsAllMonthsExcel(
    @Headers('year') year?: string,
    @Headers('supplierid') supplierId?: string,
    @Headers('routeid') routeId?: string,
    @Headers('month') month?: string,
  ) {
    return this.service.accountsAllMonthsExcel({
      year: Number(year) || undefined,
      supplierId: Number(supplierId) || undefined,
      routeId: Number(routeId) || undefined,
      month: Number(month) || undefined,
    });
  }

  @Get('AccountsAllRoundsForSupplier')
  accountsAllRounds(
    @Headers('year') year?: string,
    @Headers('month') month?: string,
    @Headers('supplierid') supplierId?: string,
    @Headers('routeid') routeId?: string,
    @Headers('pageno') pageNo?: string,
    @Headers('noofitems') noOfItems?: string,
  ) {
    return this.service.accountsAllRounds(Number(pageNo) || 1, Number(noOfItems) || 20, {
      year: Number(year) || undefined,
      month: Number(month) || undefined,
      supplierId: Number(supplierId) || undefined,
      routeId: Number(routeId) || undefined,
    });
  }

  @Get('getAllSupplierPayment')
  getAllSupplierPayment(
    @Headers('pageno') pageNo?: string,
    @Headers('noofitems') noOfItems?: string,
    @Headers('fromdate') fromDate?: string,
    @Headers('todate') toDate?: string,
    @Headers('supplierid') supplierId?: string,
  ) {
    return this.service.getAllSupplierPayment(Number(pageNo) || 1, Number(noOfItems) || 20, {
      fromDate: fromDate || undefined,
      toDate: toDate || undefined,
      supplierId: Number(supplierId) || undefined,
    });
  }

  @Post('AddSupplierPayment')
  addSupplierPayment(@Body() body: SupplierPaymentBody) {
    return this.service.addSupplierPayment(body);
  }
}

// Shape of the addSupplier payload (documented PascalCase keys).
export interface SupplierBody {
  Name?: string;
  Email?: string;
  Phone?: string;
  Mobile?: string;
  Fax?: string;
  Address?: string;
  // Contact persons: new ones have no Id, existing ones carry their Id.
  Contacts?: Array<{ Id?: number; Name?: string; Mobile?: string }>;
}

// Shape of the AddSupplierPayment payload (documented PascalCase keys, §6.12).
export interface SupplierPaymentBody {
  SupplierId?: number;
  Payment?: number;
  DatePayment?: string;
  StartDate?: string;
  TypeOfDebt?: string; // Normal | Advance
  NumberOfMonths?: number;
}

File: backend-dotnet/GarasTransport.Api/Modules/Transportation/Suppliers/SuppliersController.cs
using GarasTransport.Api.Common;
using Microsoft.AspNetCore.Mvc;

namespace GarasTransport.Api.Modules.Transportation.Suppliers;

// Base path /api/Transportation (§6.12). Every action requires CompanyName +
// UserToken headers via the filter.
[Route("api/Transportation")]
[HeaderAuth]
public class SuppliersController : ApiControllerBase
{
    private readonly SuppliersService _service;

    public SuppliersController(SuppliersService service) => _service = service;

    [HttpGet("getSuppliers")]
    public Task<object> GetSuppliers() =>
        _service.GetSuppliers(HeaderInt("PageNo", 1), HeaderInt("NoOfItems", 20), HeaderDecoded("Name"), Header("Phone"), Header("Mobile"));

    [HttpPost("addSupplier")]
    public Task<object> AddSupplier([FromBody] SupplierBody body) => _service.AddSupplier(body ?? new());

    [HttpGet("getSupplier")]
    public Task<object> GetSupplier() => _service.GetSupplier(HeaderInt("Id", 0));

    [HttpPost("updateSupplier")]
    public Task<object> UpdateSupplier([FromBody] SupplierBody body) => _service.UpdateSupplier(body?.Id ?? 0, body ?? new());

    private AccountFilters AccountFilters() => new()
    {
        Year = HeaderIntOrNull("Year"),
        Month = HeaderIntOrNull("Month"),
        SupplierId = HeaderIntOrNull("SupplierId"),
        RouteId = HeaderIntOrNull("RouteId"),
    };

    [HttpGet("AccountsAllMonthsForSupplier")]
    public Task<object> AccountsAllMonths() =>
        _service.AccountsAllMonths(HeaderInt("PageNo", 1), HeaderInt("NoOfItems", 20), AccountFilters());

    [HttpGet("AccountsAllMonthsForSupplierExcel")]
    public Task<object> AccountsAllMonthsExcel() => _service.AccountsAllMonthsExcel(AccountFilters());

    [HttpGet("AccountsAllRoundsForSupplier")]
    public Task<object> AccountsAllRounds() =>
        _service.AccountsAllRounds(HeaderInt("PageNo", 1), HeaderInt("NoOfItems", 20), AccountFilters());

    [HttpGet("getAllSupplierPayment")]
    public Task<object> GetAllSupplierPayment() =>
        _service.GetAllSupplierPayment(HeaderInt("PageNo", 1), HeaderInt("NoOfItems", 20), Header("FromDate"), Header("ToDate"), HeaderIntOrNull("SupplierId"));

    [HttpPost("AddSupplierPayment")]
    public Task<object> AddSupplierPayment([FromBody] SupplierPaymentBody body) => _service.AddSupplierPayment(body ?? new());
}
